docs/routing.md

```diff
@@ -0,0 +1,248 @@
+  - Check each source route in a phase to see if it matches.
+  - If a match is found, apply any relevant response modifiers.
+  - Determine what to do next; return the match, or continue through routing.
+- Run the matched function, or fetch the matched static asset.
+- Return the final response.
+
+### Checking Phases
+
+The first thing that happens when a request is received, is that the router starts looking for matches in the `none` phase. It calls a function to start checking phases. This function is called recursively to check additional phases.
+
+For every single source route in a phase, it calls to check whether the route is a match, and process the relevant details (more on checking source routes can be seen below). When it receives a response from the route checker, if it encountered an error — normally from running middleware — it will break out of the routing process and return the error response. Otherwise, the only other time it disables the continuation of routing is if the route checker returns a final match, i.e. it is done for that phase.
+
+After it has finished checking all the source routes in a phase, it moves on to determine what will happen next. If the current phase is the `hit` phase, or the route has at some point been rewritten to an URL, it returns that routing is complete.
+
+If the current phase is `miss`, it will update the status code to 404 if there is no file found in the build output. Afterwards, if a file was found in the output, or the current phase is `error`, it moves on to the `hit` phase. It is important that the hit phase is run every time a match is found in the build output or when the current phase is `miss` or `error`, because this allows us to update the headers for the response with the current matched path.
+
+In all other cases, it attempts to retrieve the next phase, based on the current phase. Afterwards, it recursively calls the function so that it can check the next phase.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkPhase --for each route--> checkRoute
+    checkRoute --> CheckRouteResult{check result}
+    CheckRouteResult -->|error| checkPhase
+    CheckRouteResult -->|done| IsPhaseHit{phase == hit}
+    CheckRouteResult -->|skip or next| CheckRouteResult
+    IsPhaseHit --> |yes| EndIsPhaseHit[matching done]
+    IsPhaseHit --> |no| IsPhaseMiss{phase == miss}
+    IsPhaseMiss --> |yes| Set404[set 404]
+    Set404 --> NextPhaseIsWhat{path found or phase == miss/error}
+    IsPhaseMiss --> |no| NextPhaseIsWhat
+    NextPhaseIsWhat --> |true - set next phase to hit| checkPhase
+    NextPhaseIsWhat --> |get next phase| getNextPhase --> checkPhase
+
+    getNextPhase --> NextPhaseIs{determine next phase}
+    NextPhaseIs --> |none| NextIsFS[filesystem]
+    NextPhaseIs --> |filesystem| NextIsRW[rewrite]
+    NextPhaseIs --> |rewrite| NextIsRS[resource]
+    NextPhaseIs --> |resource| NextIsMS[miss]
+    NextPhaseIs --> |all others| NextIsMS
+```
```

Comment:
Personally, I am not a fan of how much vertical space it takes up. It would be nice to be able to move the position of things and condense them, but that didn't appear to be an option.

Reply:
yeah I totally agree, as I said I did try some stuff but it's a bit tricky, I'll try to think of some alternatives :confused: